File: dev/patrol_axe/android/src/main/kotlin/pl/leancode/patrol/axe/AxeServerExtension.kt
package pl.leancode.patrol.axe

import com.google.gson.Gson
import org.http4k.core.Method.POST
import org.http4k.core.Request
import org.http4k.core.Response
import org.http4k.core.Status.Companion.BAD_REQUEST
import org.http4k.core.Status.Companion.OK
import org.http4k.routing.RoutingHttpHandler
import org.http4k.routing.bind
import org.http4k.routing.routes
import pl.leancode.patrol.PatrolServerExtension

data class AxeInitSessionRequest(val dequeApiKey: String, val dequeProjectId: String)
data class AxeScanRequest(val uploadToDashboard: Boolean, val tags: Set<String>, val scanName: String?)
data class AxeIgnoreRulesRequest(val rulesToIgnore: List<String>)
data class AxeIgnoreByViewIdResourceNameRequest(
    val viewIdResourceName: String,
    val ruleList: List<String>,
)

class AxeServerExtension : PatrolServerExtension {
    override val name: String = "patrol_axe"

    private val json = Gson()
    private val automator = AxeAutomator()

    private fun badRequest(message: String): Response = Response(BAD_REQUEST).body(message)

    private fun <T> parseRequest(req: Request, clazz: Class<T>): T? = try {
        json.fromJson(req.bodyString(), clazz)
    } catch (_: Throwable) {
        null
    }

    override fun routes(): RoutingHttpHandler = routes(
        "axeInitSession" bind POST to { req ->
            val body = parseRequest(req, AxeInitSessionRequest::class.java)
                ?: return@to badRequest("Invalid axeInitSession payload")
            if (body.dequeApiKey.isBlank() || body.dequeProjectId.isBlank()) {
                return@to badRequest("dequeApiKey and dequeProjectId are required")
            }
            automator.initSession(body.dequeApiKey, body.dequeProjectId)
            Response(OK)
        },
        "axeScan" bind POST to { req ->
            val body = parseRequest(req, AxeScanRequest::class.java)
                ?: return@to badRequest("Invalid axeScan payload")
            automator.scan(body.uploadToDashboard, body.tags, body.scanName)
            Response(OK)
        },
        "axeIgnoreRules" bind POST to { req ->
            val body = parseRequest(req, AxeIgnoreRulesRequest::class.java)
                ?: return@to badRequest("Invalid axeIgnoreRules payload")
            automator.ignoreRules(body.rulesToIgnore)
            Response(OK)
        },
        "axeIgnoreByViewIdResourceName" bind POST to { req ->
            val body = parseRequest(req, AxeIgnoreByViewIdResourceNameRequest::class.java)
                ?: return@to badRequest("Invalid axeIgnoreByViewIdResourceName payload")
            automator.ignoreByViewIdResourceName(body.viewIdResourceName, body.ruleList)
            Response(OK)
        },
        "axeIgnoreExperimental" bind POST to {
            automator.ignoreExperimental()
            Response(OK)
        },
    )
}

File: dev/patrol_axe/darwin/patrol_axe/Sources/PatrolAxe/AxeServerExtension.swift
#if PATROL_ENABLED

  import Foundation

  #if canImport(PatrolImpl)
    import PatrolImpl
  #else
    import patrol
  #endif

  @objc(AxeServerExtension)
  public class AxeServerExtension: NSObject, PatrolServerExtension {
    public override required init() { super.init() }

    public var name: String { "patrol_axe" }

    public func register(on registrar: PatrolRouteRegistrar) {
      registrar.post("axeInitSession") { body in
        let req = try JSONDecoder().decode(AxeInitSessionRequest.self, from: body)
        try AxeAutomator.shared.initSession(apiKey: req.dequeApiKey, projectId: req.dequeProjectId)
        return Data()
      }

      registrar.post("axeScan") { body in
        let req = try JSONDecoder().decode(AxeScanRequest.self, from: body)
        try AxeAutomator.shared.scan(
          uploadToDashboard: req.uploadToDashboard,
          tags: req.tags,
          scanName: req.scanName
        )
        return Data()
      }

      registrar.post("axeIgnoreRules") { body in
        let req = try JSONDecoder().decode(AxeIgnoreRulesRequest.self, from: body)
        AxeAutomator.shared.ignoreRules(req.rulesToIgnore)
        return Data()
      }

      registrar.post("axeIgnoreByViewIdResourceName") { body in
        let req = try JSONDecoder().decode(AxeIgnoreByViewIdResourceNameRequest.self, from: body)
        AxeAutomator.shared.ignoreByViewIdResourceName(
          viewIdResourceName: req.viewIdResourceName,
          ruleList: req.ruleList
        )
        return Data()
      }

      registrar.post("axeIgnoreExperimental") { _ in
        AxeAutomator.shared.ignoreExperimental()
        return Data()
      }
    }
  }

  private struct AxeInitSessionRequest: Decodable {
    let dequeApiKey: String
    let dequeProjectId: String
  }

  private struct AxeScanRequest: Decodable {
    let uploadToDashboard: Bool
    let tags: [String]
    let scanName: String?
  }

  private struct AxeIgnoreRulesRequest: Decodable {
    let rulesToIgnore: [String]
  }

  private struct AxeIgnoreByViewIdResourceNameRequest: Decodable {
    let viewIdResourceName: String
    let ruleList: [String]
  }

#endif
